gc/base/EvacuatorCopyspace.hpp

```diff
@@ -0,0 +1,199 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef EVACUATORCOPYSPACE_HPP_
+#define EVACUATORCOPYSPACE_HPP_
+
+#include "Base.hpp"
+#include "EvacuatorBase.hpp"
+#include "EvacuatorWhitelist.hpp"
+#include "SlotObject.hpp"
+
+class MM_EvacuatorCopyspace : public MM_Base
+{
+/*
+ * Data members
+ */
+private:
+
+protected:
+	uint8_t *_base;			/* points to start of copyspace */
+	uint8_t *_copy;			/* points to current copy head */
+	uint8_t *_end;			/* points to end of copyspace */
+	uintptr_t _flags;		/* extensible bitmap of flags */
+
+	/* enumeration of flag bits may be extended past copyEndFlag by subclasses */
+	typedef enum copyspaceFlags {
+		isLOAFlag = 1				/* set if base is in the large object area (LOA) of the heap */
+		, copyEndFlag = isLOAFlag	/* marks end of copyspace flags, subclasses can extend flag bit enumeration from here */
+	} copyspaceFlags;
+
+public:
+/*
+ * Function members
+ */
+private:
+
+protected:
+
+public:
+	/**
+	 * Basic array constructor obviates need for stdlibc++ linkage in gc component libraries. Array
+	 * is allocated from forge as contiguous block sized to contain requested number of elements and
+	 * must be freed using MM_Forge::free() when no longer needed.
+	 *
+	 * @param count the number of aray elements to instantiate
+	 * @return a pointer to instantiated array
+	 */
+	static MM_EvacuatorCopyspace *
+	newInstanceArray(MM_Forge *forge, uintptr_t count)
+	{
+		MM_EvacuatorCopyspace *copyspace = (MM_EvacuatorCopyspace *)forge->allocate(sizeof(MM_EvacuatorCopyspace) * count, OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
```

Comment:
Where is this memory freed?

Reply:
In MM_Evacuator::tearDown(). Otherwise MM_EvacuatorCopyspace would be required to cache a pointer to MM_Forge.

Reply:
Could this be placed in a comment for the newInstanceArray so it is clear to everyone where it is being freed.

Reply:
Will do.